### DISCUSSION TITLE
SetProgramArgs() modifies sys.argv

### DISCUSSION THREAD

**Discussion by LennertP** [pyscripter/python4delphi#430] [Q&A] [answered]
Hi!

Recently I noticed that sys.argv is modified when extending Python using PythonEngine.pas. I found that this is because SetProgramArgs() inserts an empty string in the call to PySys_SetArgv(). My understanding is that calling PySys_SetArgv() is useful when embedding Python, but when extending Python by using PythonEngine.pas to create a .pyd this seems undesirable.

I created this small pargs.py file to reproduce the issue:
import sys
print(sys.argv)

Initially the output running 'python pargs.py' is:
['pargs.py']

When I also import my own package implemented using PythonEngine.pas like this:
import mypackage
import sys
print(sys.argv)

running 'python pargs.py' again now results in;
['', 'pargs.py']

Do you think it is a good idea to skip calling the SetProgramArgs procedure when extending Python? If yes, how can we discriminate between extending Python and embedding Python in TPythonEngine.Initialize?

Best regards and many thanks for maintaining this excellent PythonEngine,

Lennert.

**Comment by pyscripter** [comment]
You are most likely not using LoadDllInExtensionModule to load the python dll in your Delphi created python module.
See https://github.com/pyscripter/python4delphi/blob/master/Modules/DemoModule/uMain.pas

**Reply by LennertP** [reply]
I am not sure what you mean. As I am writing an extension I don't load the python DLL at all. Or do you mean something else? If needed I can try to make a small extension that does (nearly) nothing but should prove my point that sys.argv is being modified.

**Reply by pyscripter** [reply, marked as answer]
Sorry for the late reply.  I have just seen your response.

> As I am writing an extension

Do you mean a python extension module (a pyd file that can be loaded into python)?  

If yes, please  see how LoadDllInExtensionModule is used in the [demo test module](https://github.com/pyscripter/python4delphi/blob/master/Modules/DemoModule/uMain.pas#L60).

**Reply by LennertP** [reply]
Thanks, using LoadDllInExtensionModule did the trick!
